aten/src/ATen/cpu/vec/vec512/vec512_float16.h

```diff
@@ -0,0 +1,887 @@
+  __m256i lo = _mm512_extracti32x8_epi32(a, 0);
+  __m256i hi = _mm512_extracti32x8_epi32(a, 1);
+  cvtfp16_fp32(lo, o1);
+  cvtfp16_fp32(hi, o2);
+}
+
+static inline __m512i cvtfp32_fp16(const __m512& a, const __m512& b) {
+  __m256i lo = _mm512_cvtps_ph(
+      a, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  __m256i hi = _mm512_cvtps_ph(
+      b, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  __m512 t_lo = _mm512_castsi512_ps(_mm512_castsi256_si512(lo));
+  __m256 t_hi = _mm256_castsi256_ps(hi);
+  return _mm512_castps_si512(_mm512_insertf32x8(t_lo, t_hi, 1));
+}
+
+template <> class Vectorized<Half> {
+private:
+  __m512i values;
+public:
+  using value_type = uint16_t;
+  using size_type = int;
+  static constexpr size_type size() {
+    return 32;
+  }
+  Vectorized() {}
+  Vectorized(__m512i v) : values(v) {}
+  Vectorized(Half val) {
+    value_type uw = val.x;
+    values = _mm512_set1_epi16(uw);
+  }
+  Vectorized(Half val1, Half val2, Half val3, Half val4,
+         Half val5, Half val6, Half val7, Half val8,
+         Half val9, Half val10, Half val11, Half val12,
+         Half val13, Half val14, Half val15, Half val16,
+         Half val17, Half val18, Half val19, Half val20,
+         Half val21, Half val22, Half val23, Half val24,
+         Half val25, Half val26, Half val27, Half val28,
+         Half val29, Half val30, Half val31, Half val32) {
+    values = _mm512_set_epi16(
+        val32.x, val31.x, val30.x, val29.x, val28.x, val27.x, val26.x, val25.x,
+        val24.x, val23.x, val22.x, val21.x, val20.x, val19.x, val18.x, val17.x,
+        val16.x, val15.x, val14.x, val13.x, val12.x, val11.x, val10.x, val9.x,
+        val8.x, val7.x, val6.x, val5.x, val4.x, val3.x, val2.x, val1.x);
+  }
+  operator __m512i() const {
+    return values;
+  }
+  Half& operator[](int idx) = delete;
+  const Half& operator[](int idx) const  = delete;
+  int zero_mask() const {
+    // returns an integer mask where all zero elements are translated to 1-bit and others are translated to 0-bit
+    return _mm512_cmpeq_epi16_mask(values, _mm512_set1_epi16(0));
+  }
+  static Vectorized<Half> loadu(const void* ptr) {
+    return _mm512_loadu_si512(reinterpret_cast<const __m512i*>(ptr));
+  }
+  static Vectorized<Half> loadu(const void* ptr, int16_t count) {
+    __at_align__ int16_t tmp_values[size()];
+    std::memcpy(tmp_values, ptr, count * sizeof(int16_t));
+    return loadu(tmp_values);
+  }
+  void store(void* ptr, int count = size()) const {
+    if (count == size()) {
+      _mm512_storeu_si512(reinterpret_cast<__m512i*>(ptr), values);
+    } else if (count > 0) {
+      __at_align__ int16_t tmp_values[size()];
+      _mm512_storeu_si512(reinterpret_cast<__m512i*>(tmp_values), values);
+      std::memcpy(ptr, tmp_values, count * sizeof(int16_t));
+    }
+  }
+  template <int64_t mask>
+  static Vectorized<Half> blend(const Vectorized<Half>& a, const Vectorized<Half>& b) {
```

Comment:
This looks like exact duplicate of https://github.com/pytorch/pytorch/blob/191a2322f01cf8646a916bdbdcdbfb352a9cc263/aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h#L122-L126

Can it be avoided somehow?

Reply:
Not only this function, but for others, can we share the implementation between bf16 and fp16 somehow, e.g., create a common base class. There are three kinds of functions:
1. Share exact same implementation, which can be put into the base.
2. Share the main fp32 compute but with conversion ops in the prefix/suffix. The main compute body can be put into the base.
3. Different implementation, put as overrides in the subclasses.

Reply:
I have abstracted the common code to `struct Vectorized16` to reduce the duplicate code between `Vectorized<BFloat16>` and `Vectorized<Half>`. And used some pre processor on some occasions.